File: android/domain/src/main/kotlin/com/bolsillo/domain/usecase/SoftDeleteTransaction.kt
package com.bolsillo.domain.usecase

import com.bolsillo.domain.port.TransactionRepository
import javax.inject.Inject

/**
 * FR 12 — soft delete a single transaction, or the whole transfer group if
 * the row carries a [transferGroupId]. Article III: no hard deletes.
 */
class SoftDeleteTransaction
    @Inject
    constructor(private val transactions: TransactionRepository) {
        suspend operator fun invoke(
            id: String,
            now: Long = System.currentTimeMillis(),
        ) {
            val target = transactions.getById(id) ?: return
            val group = target.transferGroupId
            if (group != null) {
                transactions.softDeleteGroup(group, now)
            } else {
                transactions.softDelete(id, now)
            }
        }
    }

File: ios/Packages/Bolsillo/Sources/BolsilloDomain/UseCases/SoftDeleteTransaction.swift
import Foundation

/// Soft-deletes a transaction (moves it to trash). If the transaction belongs to a
/// transfer group, both legs are soft-deleted atomically to preserve the double-entry
/// invariant (Constitution Article III — Invariants 6 & 7).
///
/// There are NO hard deletes (Constitution Article III).
public struct SoftDeleteTransaction: Sendable {
    private let transactions: any TransactionRepository

    public init(transactions: some TransactionRepository) {
        self.transactions = transactions
    }

    /// Soft-delete the transaction identified by `id`.
    ///
    /// - Parameters:
    ///   - id: The transaction to delete.
    ///   - now: The deletion timestamp (injected for testability).
    public func callAsFunction(id: String, now: Date) async {
        if let tx = await transactions.getById(id), let groupId = tx.transferGroupId {
            // Transfer: soft-delete the whole group atomically (Invariant 7)
            await transactions.softDeleteGroup(transferGroupId: groupId, deletedAt: now)
        } else {
            await transactions.softDelete(id: id, deletedAt: now)
        }
    }
}
